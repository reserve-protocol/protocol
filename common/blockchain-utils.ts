import { BigNumber } from 'ethers'
import { HardhatRuntimeEnvironment } from 'hardhat/types'

// getChainId: Returns current chain Id
export const getChainId = async (hre: HardhatRuntimeEnvironment) => {
  let _chainId
  try {
    _chainId = await hre.network.provider.send('eth_chainId')
  } catch (e) {
    console.log('failed to get chainId, falling back on net_version...')
    _chainId = await hre.network.provider.send('net_version')
  }

  if (!_chainId) {
    throw new Error(`could not get chainId from network`)
  }
  if (_chainId.startsWith('0x')) {
    _chainId = BigNumber.from(_chainId).toString()
  }
  return _chainId
}
