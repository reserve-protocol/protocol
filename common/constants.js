const { BigNumber } = require("ethers");

const ZERO_ADDRESS = '0x0000000000000000000000000000000000000000';

const ONE_ETH= BigNumber.from("1000000000000000000");

const MAX_UINT256 = BigNumber.from(2).pow(256).sub(1);
const MAX_UINT16 = (2**16) - 1;

const SCALE_FACTOR = 1e18;
const BN_SCALE_FACTOR = BigNumber.from(SCALE_FACTOR.toString());

module.exports = {
    ZERO_ADDRESS,
    MAX_UINT256,
    MAX_UINT16,
    SCALE_FACTOR,
    BN_SCALE_FACTOR,
    ONE_ETH
};
