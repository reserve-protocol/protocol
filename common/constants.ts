import { ethers } from 'hardhat'
import { BigNumber } from 'ethers'

export const ZERO_ADDRESS = '0x0000000000000000000000000000000000000000'
export const ONE_ADDRESS = '0x0000000000000000000000000000000000000001'

export const ONE_ETH = BigNumber.from('1000000000000000000')

export const MAX_UINT256 = BigNumber.from(2).pow(256).sub(1)
export const MAX_UINT192 = BigNumber.from(2).pow(192).sub(1)
export const MAX_UINT96 = BigNumber.from(2).pow(96).sub(1)
export const MAX_UINT32 = 2 ** 32 - 1
export const MAX_UINT16 = 2 ** 16 - 1

export const SCALE_DECIMALS = 18
export const SCALE_FACTOR = 10 ** SCALE_DECIMALS
export const BN_SCALE_FACTOR = BigNumber.from(SCALE_FACTOR.toString())

// @dev Must match `IAsset.CollateralStatus`.
export enum CollateralStatus {
  SOUND,
  IFFY,
  DISABLED,
}

// @dev Must match `Governance.ProposalState`.
export enum ProposalState {
  Pending,
  Active,
  Canceled,
  Defeated,
  Succeeded,
  Queued,
  Expired,
  Executed,
}

// @dev Must match `Fixed.RoundingApproach`.
export enum RoundingMode {
  FLOOR,
  ROUND,
  CEIL,
}

// @dev Must match `GnosisTrade.TradeStatus`.
export enum TradeStatus {
  NOT_STARTED,
  OPEN,
  CLOSED,
}

export const FURNACE_DEST = '0x0000000000000000000000000000000000000001'
export const STRSR_DEST = '0x0000000000000000000000000000000000000002'

export const QUEUE_START = '0x0000000000000000000000000000000000000000000000000000000000000001'

// Auth roles
export const OWNER = ethers.utils.formatBytes32String('OWNER')
export const SHORT_FREEZER = ethers.utils.formatBytes32String('SHORT_FREEZER')
export const LONG_FREEZER = ethers.utils.formatBytes32String('LONG_FREEZER')
export const PAUSER = ethers.utils.formatBytes32String('PAUSER')
