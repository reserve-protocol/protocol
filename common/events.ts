import { expect } from 'chai'
import { BigNumber, ContractReceipt, Event } from 'ethers'

// TODO: Proper typing
const contains = (args: { [key: string]: any | undefined }, key: string, value: any): any => {
  expect(key in args).to.equal(true, `Event argument '${key}' not found`)

  if (value === null) {
    expect(args[key]).to.equal(null, `expected event argument '${key}' to be null but got ${args[key]}`)
  } else if (BigNumber.isBigNumber(args[key]) || BigNumber.isBigNumber(value)) {
    const actual = BigNumber.isBigNumber(args[key]) ? args[key].toString() : args[key]
    const expected = BigNumber.isBigNumber(value) ? value.toString() : value

    expect(args[key]).to.equal(value, `expected event argument '${key}' to have value ${expected} but got ${actual}`)
  } else {
    expect(args[key]).to.be.deep.equal(
      value,
      `expected event argument '${key}' to have value ${value} but got ${args[key]}`
    )
  }
}

// TODO: Proper typing for "eventArgs"
export const expectInReceipt = (receipt: ContractReceipt, eventName: string, eventArgs = {}): any => {
  if (receipt.events == undefined) {
    throw new Error('No events found in receipt')
  }

  const events = receipt.events.filter((e: Event) => e.event === eventName)
  expect(events.length > 0).to.equal(true, `No '${eventName}' events found`)

  const exceptions: string[] = []
  const event = events.find(function (e: Event) {
    for (const [k, v] of Object.entries(eventArgs)) {
      try {
        if (e.args == undefined) {
          throw new Error('Event has no arguments')
        }

        contains(e.args, k, v)
      } catch (error) {
        exceptions.push(error as string)
        return false
      }
    }
    return true
  })

  if (event === undefined) {
    // Each event entry may have failed to match for different reasons,
    // throw the first one
    throw exceptions[0]
  }

  return event
}
