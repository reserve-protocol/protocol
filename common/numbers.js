const { BigNumber } = require("ethers");
const { BN_SCALE_FACTOR } = require("./constants");

const bn = (x) => {
    if (BigNumber.isBigNumber(x)) return x;
    const stringified = parseScientific(x.toString());
    const integer = stringified.split('.')[0];
    return BigNumber.from(integer);
};

const fp = (x) => {
    return BN_SCALE_FACTOR.mul(x);
}
 
function parseScientific(num) {
    // If the number is not in scientific notation return it as it is
    if (!/\d+\.?\d*e[+-]*\d+/i.test(num)) return num;

    // Remove the sign
    const numberSign = Math.sign(Number(num));
    num = Math.abs(Number(num)).toString();

    // Parse into coefficient and exponent
    const [coefficient, exponent] = num.toLowerCase().split('e');
    let zeros = Math.abs(Number(exponent));
    const exponentSign = Math.sign(Number(exponent));
    const [integer, decimals] = (coefficient.indexOf('.') != -1 ? coefficient : `${coefficient}.`).split('.');

    if (exponentSign === -1) {
        zeros -= integer.length;
        num =
            zeros < 0
                ? integer.slice(0, zeros) + '.' + integer.slice(zeros) + decimals
                : '0.' + '0'.repeat(zeros) + integer + decimals;
    } else {
        if (decimals) zeros -= decimals.length;
        num =
            zeros < 0
                ? integer + decimals.slice(0, zeros) + '.' + decimals.slice(zeros)
                : integer + decimals + '0'.repeat(zeros);
    }

    return numberSign < 0 ? '-' + num : num;
}

module.exports = {
    bn,
    fp
}