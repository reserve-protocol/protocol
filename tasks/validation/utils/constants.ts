import { networkConfig } from '#/common/configuration'

export const whales: { [key: string]: string } = {
  [networkConfig['1'].tokens.USDT!.toLowerCase()]: '0x47ac0Fb4F2D84898e4D9E7b4DaB3C24507a6D503',
  [networkConfig['1'].tokens.USDC!.toLowerCase()]: '0xAFAaDfa18D9d63d09F19a5445e29CEc601054C5e',
  [networkConfig['1'].tokens.pyUSD!.toLowerCase()]: '0xA5588F7cdf560811710A2D82D3C9c99769DB1Dcb',
  [networkConfig['1'].tokens.RSR!.toLowerCase()]: '0x6bab6EB87Aa5a1e4A8310C73bDAAA8A5dAAd81C1',
  [networkConfig['1'].tokens.cUSDT!.toLowerCase()]: '0xb99CC7e10Fe0Acc68C50C7829F473d81e23249cc',
  [networkConfig['1'].tokens.aUSDT!.toLowerCase()]: '0x0B6B712B0f3998961Cd3109341b00c905b16124A',
  ['0x21fe646D1Ed0733336F2D4d9b2FE67790a6099D9'.toLowerCase()]:
    '0x5754284f345afc66a98fbB0a0Afe71e0F007B949', // saUSDT
  ['0x4Be33630F92661afD646081BC29079A38b879aA0'.toLowerCase()]:
    '0x5754284f345afc66a98fbB0a0Afe71e0F007B949', // cUSDTVault

  [networkConfig['1'].tokens.aUSDC!.toLowerCase()]: '0x777777c9898D384F785Ee44Acfe945efDFf5f3E0',
  [networkConfig['1'].tokens.cUSDC!.toLowerCase()]: '0x97D868b5C2937355Bf89C5E5463d52016240fE86',
  [networkConfig['1'].tokens.cUSDCv3!.toLowerCase()]: '0x7f714b13249BeD8fdE2ef3FBDfB18Ed525544B03',
  ['0x60C384e226b120d93f3e0F4C502957b2B9C32B15'.toLowerCase()]:
    '0x51eDF02152EBfb338e03E30d65C15fBf06cc9ECC', // saUSDC
  ['0xf579F9885f1AEa0d3F8bE0F18AfED28c92a43022'.toLowerCase()]:
    '0x51eDF02152EBfb338e03E30d65C15fBf06cc9ECC', // cUSDCVault
  [networkConfig['1'].tokens.RSR!.toLowerCase()]: '0x6bab6EB87Aa5a1e4A8310C73bDAAA8A5dAAd81C1',
  [networkConfig['1'].tokens.WBTC!.toLowerCase()]: '0x8eb8a3b98659cce290402893d0123abb75e3ab28',
  [networkConfig['1'].tokens.stETH!.toLowerCase()]: '0x176F3DAb24a159341c0509bB36B833E7fdd0a132',
  [networkConfig['1'].tokens.WETH!.toLowerCase()]: '0x8EB8a3b98659Cce290402893d0123abb75E3ab28',
  [networkConfig['1'].tokens.DAI!.toLowerCase()]: '0x8EB8a3b98659Cce290402893d0123abb75E3ab28',
  [networkConfig['1'].tokens.CRV!.toLowerCase()]: '0xf977814e90da44bfa03b6295a0616a897441acec',
  ['0xAEda92e6A3B1028edc139A4ae56Ec881f3064D4F'.toLowerCase()]:
    '0x8605dc0C339a2e7e85EEA043bD29d42DA2c6D784', // cvxeUSDFRAXBP LP token
  [networkConfig['1'].tokens.sDAI!.toLowerCase()]: '0x4aa42145Aa6Ebf72e164C9bBC74fbD3788045016',
  ['0xa0d69e286b938e21cbf7e51d71f6a4c8918f482f'.toLowerCase()]:
    '0x3154Cf16ccdb4C6d922629664174b904d80F2C35', // eUSD
  ['0xacdf0dba4b9839b96221a8487e9ca660a48212be'.toLowerCase()]:
    '0x8a8434A5952aC2CF4927bbEa3ace255c6dd165CD', // hyUSD
}

export const collateralToUnderlying: { [key: string]: string } = {
  ['0x21fe646D1Ed0733336F2D4d9b2FE67790a6099D9'.toLowerCase()]:
    networkConfig['1'].tokens.USDT!.toLowerCase(),
  ['0x60C384e226b120d93f3e0F4C502957b2B9C32B15'.toLowerCase()]:
    networkConfig['1'].tokens.USDC!.toLowerCase(),
  [networkConfig['1'].tokens.cUSDT!.toLowerCase()]: networkConfig['1'].tokens.USDT!.toLowerCase(),
  [networkConfig['1'].tokens.cUSDC!.toLowerCase()]: networkConfig['1'].tokens.USDC!.toLowerCase(),
  [networkConfig['1'].tokens.saEthUSDC!.toLowerCase()]:
    networkConfig['1'].tokens.aEthUSDC!.toLowerCase(),
}

export interface RTokenDeployment {
  rToken: string
  governor: string
  timelock: string
}

export const MAINNET_DEPLOYMENTS: RTokenDeployment[] = [
  {
    rToken: '0xA0d69E286B938e21CBf7E51D71F6A4c8918f482F', // eUSD
    governor: '0x7e880d8bD9c9612D6A9759F96aCD23df4A4650E6',
    timelock: '0xc8Ee187A5e5c9dC9b42414Ddf861FFc615446a2c',
  },
  {
    rToken: '0xE72B141DF173b999AE7c1aDcbF60Cc9833Ce56a8', // ETH+
    governor: '0x239cDcBE174B4728c870A24F77540dAB3dC5F981',
    timelock: '0x5f4A10aE2fF68bE3cdA7d7FB432b10C6BFA6457B',
  },
  {
    rToken: '0xaCdf0DBA4B9839b96221a8487e9ca660a48212be', // hyUSD (mainnet)
    governor: '0x22d7937438b4bBf02f6cA55E3831ABB94Bd0b6f1',
    timelock: '0x624f9f076ED42ba3B37C3011dC5a1761C2209E1C',
  },
  {
    rToken: '0xFc0B1EEf20e4c68B3DCF36c4537Cfa7Ce46CA70b', // USDC+
    governor: '0xc837C557071D604bCb1058c8c4891ddBe8FDD630',
    timelock: '0x6C957417cB6DF6e821eec8555DEE8b116C291999',
  },
  {
    rToken: '0x0d86883FAf4FfD7aEb116390af37746F45b6f378', // USD3
    governor: '0x020CB71181008369C388CaAEE98b0E69f8F4C471',
    timelock: '0xE0289984F709fc7150E646B672bfaDC879a15f14',
  },
  {
    rToken: '0x78da5799CF427Fee11e9996982F4150eCe7a99A7', // rgUSD
    governor: '0x409bAc94c4207C6627EA5f4E4FFB7128e8F654Fc',
    timelock: '0x9aD9E73e38c8506a664A3A37e8A9CE910B6FBeb4',
  },
]

export const BASE_DEPLOYMENTS: RTokenDeployment[] = [
  {
    rToken: '0xCc7FF230365bD730eE4B352cC2492CEdAC49383e', // hyUSD (base)
    governor: '0xc8e63d3501A246fa1ddBAbe4ad0B50e9d32aA8bb',
    timelock: '0xf093d7f00f3dCe6d415Be564f41Cb4bc032fb367',
  },
  {
    rToken: '0xCb327b99fF831bF8223cCEd12B1338FF3aA322Ff', // bsdETH
    governor: '0xB05C6a7242595f2E23CC6a0aB20699d63D0939Fd',
    timelock: '0x321f7493B8B675dFfE2570Bd0F164237D445b9E8',
  },
  {
    rToken: '0xfE0D6D83033e313691E96909d2188C150b834285', // iUSDC - Assets skipped (USDbC)
    governor: '0xfe637F7D5B848392c19052631d68F8AC859F71cF',
    timelock: '0xd18ED37CA912bbf1EDE93d27459d03DC4343dea1',
  },
  {
    rToken: '0xC9a3e2B3064c1c0546D3D0edc0A748E9f93Cf18d', // Vaya - Assets skipped (USDbC)
    governor: '0xEb583EA06501f92E994C353aD2741A35582987aA',
    timelock: '0xeE3eC997A37e661a42673D7A489Fbf0E5ed0C223',
  },
  {
    rToken: '0x641B0453487C9D14c5df96d45a481ef1dc84e31f', // MAAT
    governor: '0x0f7f1442dA7F687BB877Fbee0539FA8D6e4d1a02',
    timelock: '0xE67cEb03EfdF9B3fb5C3FeBF3103e2efd3a76A1b',
  },
]
