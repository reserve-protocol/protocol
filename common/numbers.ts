import { BigNumber } from 'ethers'
import { ethers } from 'hardhat'
import { BN_SCALE_FACTOR, SCALE_DECIMALS } from './constants'

export const ZERO = BigNumber.from(0)

export const bn = (x: BigNumber | number): BigNumber => {
  if (BigNumber.isBigNumber(x)) return x
  const stringified = parseScientific(x.toString())
  const integer = stringified.split('.')[0]
  return BigNumber.from(integer)
}

export const pow10 = (exponent: number): BigNumber => {
  return bn(10).pow(exponent)
}

export const fp = (x: BigNumber | number): BigNumber => {
  if (BigNumber.isBigNumber(x)) {
    return BN_SCALE_FACTOR.mul(x)
  }
  return ethers.utils.parseUnits(x.toString(), SCALE_DECIMALS)
}

export const divCeil = (x: BigNumber, y: BigNumber): BigNumber =>
  // ceil(x/y) == (x + y - 1) / y
  x.add(y).sub(1).div(y)

function parseScientific(num: string) {
  // If the number is not in scientific notation return it as it is
  if (!/\d+\.?\d*e[+-]*\d+/i.test(num)) return num

  // Remove the sign
  const numberSign = Math.sign(Number(num))
  num = Math.abs(Number(num)).toString()

  // Parse into coefficient and exponent
  const [coefficient, exponent] = num.toLowerCase().split('e')
  let zeros = Math.abs(Number(exponent))
  const exponentSign = Math.sign(Number(exponent))
  const [integer, decimals] = (coefficient.indexOf('.') != -1 ? coefficient : `${coefficient}.`).split('.')

  if (exponentSign === -1) {
    zeros -= integer.length
    num =
      zeros < 0
        ? integer.slice(0, zeros) + '.' + integer.slice(zeros) + decimals
        : '0.' + '0'.repeat(zeros) + integer + decimals
  } else {
    if (decimals) zeros -= decimals.length
    num =
      zeros < 0
        ? integer + decimals.slice(0, zeros) + '.' + decimals.slice(zeros)
        : integer + decimals + '0'.repeat(zeros)
  }

  return numberSign < 0 ? '-' + num : num
}
