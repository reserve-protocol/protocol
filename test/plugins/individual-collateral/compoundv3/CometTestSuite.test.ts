import collateralTests from '../collateralTests'
import {
  CollateralFixtureContext,
  CollateralOpts,
  MintCollateralFunc,
  CollateralStatus,
} from '../pluginTestTypes'
import { mintWcUSDC, makewCSUDC, resetFork, enableRewardsAccrual } from './helpers'
import { ethers } from 'hardhat'
import { ContractFactory, BigNumberish } from 'ethers'
import {
  ERC20Mock,
  MockV3Aggregator,
  CometInterface,
  ICusdcV3Wrapper,
  ICusdcV3WrapperMock,
  CusdcV3WrapperMock,
  CusdcV3Wrapper__factory,
  CusdcV3WrapperMock__factory,
  MockV3Aggregator__factory,
  CometMock,
  CometMock__factory,
  TestICollateral,
} from '../../../../typechain'
import { bn } from '../../../../common/numbers'
import { MAX_UINT48 } from '../../../../common/constants'
import { expect } from 'chai'
import { SignerWithAddress } from '@nomiclabs/hardhat-ethers/signers'
import {
  advanceTime,
  getLatestBlockTimestamp,
  setNextBlockTimestamp,
  advanceBlocks,
} from '../../../utils/time'
import {
  ORACLE_ERROR,
  ORACLE_TIMEOUT,
  COMP,
  CUSDC_V3,
  USDC_USD_PRICE_FEED,
  MAX_TRADE_VOL,
  DEFAULT_THRESHOLD,
  DELAY_UNTIL_DEFAULT,
  REWARDS,
  USDC,
} from './constants'

/*
  Define interfaces
*/

interface CometCollateralFixtureContext extends CollateralFixtureContext {
  cusdcV3: CometInterface
  wcusdcV3: ICusdcV3Wrapper
  usdc: ERC20Mock
  wcusdcV3Mock: CusdcV3WrapperMock
}

interface CometCollateralFixtureContextMockComet extends CollateralFixtureContext {
  cusdcV3: CometMock
  wcusdcV3: ICusdcV3Wrapper
  usdc: ERC20Mock
  wcusdcV3Mock: CusdcV3WrapperMock
}

interface CometCollateralOpts extends CollateralOpts {
  reservesThresholdIffy?: BigNumberish
  reservesThresholdDisabled?: BigNumberish
}

/*
  Define deployment functions
*/

const chainlinkDefaultAnswer = bn('1e8')

export const defaultCometCollateralOpts: CometCollateralOpts = {
  erc20: CUSDC_V3,
  targetName: ethers.utils.formatBytes32String('USD'),
  rewardERC20: COMP,
  priceTimeout: ORACLE_TIMEOUT,
  chainlinkFeed: USDC_USD_PRICE_FEED,
  oracleTimeout: ORACLE_TIMEOUT,
  oracleError: ORACLE_ERROR,
  maxTradeVolume: MAX_TRADE_VOL,
  defaultThreshold: DEFAULT_THRESHOLD,
  delayUntilDefault: DELAY_UNTIL_DEFAULT,
  reservesThresholdIffy: bn('10000'),
  reservesThresholdDisabled: bn('5000'),
}

export const deployCollateral = async (
  opts: CometCollateralOpts = {}
): Promise<TestICollateral> => {
  opts = { ...defaultCometCollateralOpts, ...opts }

  const CTokenV3CollateralFactory: ContractFactory = await ethers.getContractFactory(
    'CTokenV3Collateral'
  )

  const collateral = <TestICollateral>await CTokenV3CollateralFactory.deploy(
    {
      erc20: opts.erc20,
      targetName: opts.targetName,
      priceTimeout: opts.priceTimeout,
      chainlinkFeed: opts.chainlinkFeed,
      oracleError: opts.oracleError,
      oracleTimeout: opts.oracleTimeout,
      maxTradeVolume: opts.maxTradeVolume,
      defaultThreshold: opts.defaultThreshold,
      delayUntilDefault: opts.delayUntilDefault,
    },
    {
      rewardERC20: opts.rewardERC20,
      reservesThresholdIffy: opts.reservesThresholdIffy,
      reservesThresholdDisabled: opts.reservesThresholdDisabled,
    },
    0,
    { gasLimit: 2000000000 }
  )
  await collateral.deployed()

  // sometimes we are trying to test a negative test case and we want this to fail silently
  // fortunately this syntax fails silently because our tools are terrible
  await expect(collateral.refresh())

  return collateral
}

type Fixture<T> = () => Promise<T>

const makeCollateralFixtureContext = (
  alice: SignerWithAddress,
  opts: CometCollateralOpts = {}
): Fixture<CometCollateralFixtureContext> => {
  const collateralOpts = { ...defaultCometCollateralOpts, ...opts }

  const makeCollateralFixtureContext = async () => {
    const MockV3AggregatorFactory = <MockV3Aggregator__factory>(
      await ethers.getContractFactory('MockV3Aggregator')
    )

    const chainlinkFeed = <MockV3Aggregator>(
      await MockV3AggregatorFactory.deploy(8, chainlinkDefaultAnswer)
    )
    collateralOpts.chainlinkFeed = chainlinkFeed.address

    const fix = await makewCSUDC()
    const cusdcV3 = <CometInterface>fix.cusdcV3
    const { wcusdcV3, usdc } = fix

    const CusdcV3WrapperMockFactory = <CusdcV3WrapperMock__factory>(
      await ethers.getContractFactory('CusdcV3WrapperMock')
    )
    const wcusdcV3Mock = <ICusdcV3WrapperMock>(
      ((await CusdcV3WrapperMockFactory.deploy(wcusdcV3.address)) as ICusdcV3WrapperMock)
    )
    const realMock = (await ethers.getContractAt(
      'ICusdcV3WrapperMock',
      wcusdcV3Mock.address
    )) as ICusdcV3WrapperMock
    collateralOpts.erc20 = realMock.address
    const collateral = await deployCollateral(collateralOpts)
    const rewardToken = <ERC20Mock>await ethers.getContractAt('ERC20Mock', COMP)
    const tokDecimals = await wcusdcV3.decimals()

    return {
      alice,
      collateral,
      chainlinkFeed,
      cusdcV3,
      wcusdcV3: <ICusdcV3WrapperMock>realMock,
      wcusdcV3Mock,
      usdc,
      tok: wcusdcV3,
      rewardToken,
      tokDecimals,
    }
  }

  return makeCollateralFixtureContext
}

const deployCollateralCometMockContext = async (
  opts: CometCollateralOpts = {}
): Promise<CometCollateralFixtureContextMockComet> => {
  const collateralOpts = { ...defaultCometCollateralOpts, ...opts }

  const MockV3AggregatorFactory = <MockV3Aggregator__factory>(
    await ethers.getContractFactory('MockV3Aggregator')
  )
  const chainlinkFeed = <MockV3Aggregator>await MockV3AggregatorFactory.deploy(8, bn('1e8'))
  collateralOpts.chainlinkFeed = chainlinkFeed.address

  const CometFactory = <CometMock__factory>await ethers.getContractFactory('CometMock')
  const cusdcV3 = <CometMock>await CometFactory.deploy(bn('5e15'), bn('1e15'), CUSDC_V3)

  const CusdcV3WrapperFactory = <CusdcV3Wrapper__factory>(
    await ethers.getContractFactory('CusdcV3Wrapper')
  )

  const wcusdcV3 = <ICusdcV3Wrapper>(
    ((await CusdcV3WrapperFactory.deploy(
      cusdcV3.address,
      REWARDS,
      COMP
    )) as unknown as ICusdcV3Wrapper)
  )
  const CusdcV3WrapperMockFactory = <CusdcV3WrapperMock__factory>(
    await ethers.getContractFactory('CusdcV3WrapperMock')
  )
  const wcusdcV3Mock = await (<ICusdcV3WrapperMock>(
    await CusdcV3WrapperMockFactory.deploy(wcusdcV3.address)
  ))

  const realMock = (await ethers.getContractAt(
    'ICusdcV3WrapperMock',
    wcusdcV3Mock.address
  )) as ICusdcV3WrapperMock
  collateralOpts.erc20 = wcusdcV3.address
  collateralOpts.erc20 = realMock.address
  const usdc = <ERC20Mock>await ethers.getContractAt('ERC20Mock', USDC)
  const collateral = await deployCollateral(collateralOpts)

  const rewardToken = <ERC20Mock>await ethers.getContractAt('ERC20Mock', COMP)
  const tokDecimals = await wcusdcV3.decimals()

  return {
    collateral,
    chainlinkFeed,
    cusdcV3,
    wcusdcV3: <ICusdcV3WrapperMock>wcusdcV3Mock,
    wcusdcV3Mock,
    usdc,
    tok: wcusdcV3,
    rewardToken,
    tokDecimals,
  }
}

/*
  Define helper functions
*/

const mintCollateralTo: MintCollateralFunc<CometCollateralFixtureContext> = async (
  ctx: CometCollateralFixtureContext,
  amount: BigNumberish,
  user: SignerWithAddress,
  recipient: string
) => {
  await mintWcUSDC(ctx.usdc, ctx.cusdcV3, ctx.wcusdcV3, user, amount, recipient)
}

const reduceRefPerTok = async (ctx: CometCollateralFixtureContext) => {
  const currentExchangeRate = await ctx.wcusdcV3.exchangeRate()
  await ctx.wcusdcV3Mock.setMockExchangeRate(true, currentExchangeRate.sub(100))
}

const increaseRefPerTok = async () => {
  await advanceBlocks(1000)
  await setNextBlockTimestamp((await getLatestBlockTimestamp()) + 12000)
}

/*
  Define collateral-specific tests
*/

const collateralSpecificConstructorTests = () => {
  it('does not allow missing rewardERC20', async () => {
    await expect(
      deployCollateral({ rewardERC20: ethers.constants.AddressZero })
    ).to.be.revertedWith('rewardERC20 missing')
  })

  it('does not allow 0 reservesThresholdIffy', async () => {
    await expect(
      deployCollateral({ erc20: CUSDC_V3, reservesThresholdIffy: 0 })
    ).to.be.revertedWith('reservesThresholdIffy zero')
  })

  it('does not allow 0 reservesThresholdDisabled', async () => {
    await expect(
      deployCollateral({ erc20: CUSDC_V3, reservesThresholdDisabled: 0 })
    ).to.be.revertedWith('reservesThresholdDisabled zero')
  })
}

const collateralSpecificStatusTests = () => {
  it('enters IFFY state when compound reserves are below target reserves iffy threshold', async () => {
    const mockOpts = { reservesThresholdIffy: 5000n, reservesThresholdDisabled: 1000n }
    const { collateral, cusdcV3 } = await deployCollateralCometMockContext(mockOpts)
    const delayUntilDefault = await collateral.delayUntilDefault()

    // Check initial state
    await expect(collateral.refresh()).to.not.emit(collateral, 'CollateralStatusChanged')
    expect(await collateral.status()).to.equal(CollateralStatus.SOUND)
    expect(await collateral.whenDefault()).to.equal(MAX_UINT48)

    // cUSDC/Comet's reserves gone down below reservesThresholdIffy
    await cusdcV3.setReserves(4000n)

    const nextBlockTimestamp = (await getLatestBlockTimestamp()) + 1
    await setNextBlockTimestamp(nextBlockTimestamp)
    const expectedDefaultTimestamp = nextBlockTimestamp + delayUntilDefault

    await expect(collateral.refresh())
      .to.emit(collateral, 'CollateralStatusChanged')
      .withArgs(CollateralStatus.SOUND, CollateralStatus.IFFY)
    expect(await collateral.status()).to.equal(CollateralStatus.IFFY)
    expect(await collateral.whenDefault()).to.equal(expectedDefaultTimestamp)

    // Move time forward past delayUntilDefault
    await advanceTime(delayUntilDefault)
    expect(await collateral.status()).to.equal(CollateralStatus.DISABLED)

    // Nothing changes if attempt to refresh after default for CTokenV3
    const prevWhenDefault: bigint = (await collateral.whenDefault()).toBigInt()
    await expect(collateral.refresh()).to.not.emit(collateral, 'CollateralStatusChanged')
    expect(await collateral.status()).to.equal(CollateralStatus.DISABLED)
    expect(await collateral.whenDefault()).to.equal(prevWhenDefault)
  })

  it('enters DISABLED state when reserves threshold is at disabled levels', async () => {
    const mockOpts = { reservesThresholdDisabled: 1000n }
    const { collateral, cusdcV3 } = await deployCollateralCometMockContext(mockOpts)

    // Check initial state
    expect(await collateral.status()).to.equal(CollateralStatus.SOUND)
    expect(await collateral.whenDefault()).to.equal(MAX_UINT48)

    // cUSDC/Comet's reserves gone down to 19% of target reserves
    await cusdcV3.setReserves(900n)

    await expect(collateral.refresh()).to.emit(collateral, 'CollateralStatusChanged')
    // State remains the same
    expect(await collateral.status()).to.equal(CollateralStatus.DISABLED)
    expect(await collateral.whenDefault()).to.equal(await getLatestBlockTimestamp())
  })

  it('enters DISABLED state if reserves go negative', async () => {
    const mockOpts = { reservesThresholdDisabled: 1000n }
    const { collateral, cusdcV3 } = await deployCollateralCometMockContext(mockOpts)

    // Check initial state
    expect(await collateral.status()).to.equal(CollateralStatus.SOUND)
    expect(await collateral.whenDefault()).to.equal(MAX_UINT48)

    // cUSDC/Comet's reserves gone down to -1
    await cusdcV3.setReserves(-1)

    await expect(collateral.refresh()).to.emit(collateral, 'CollateralStatusChanged')
    // State remains the same
    expect(await collateral.status()).to.equal(CollateralStatus.DISABLED)
    expect(await collateral.whenDefault()).to.equal(await getLatestBlockTimestamp())
  })
}

const beforeEachRewardsTest = async (ctx: CometCollateralFixtureContext) => {
  await enableRewardsAccrual(ctx.cusdcV3)
}

/*
  Run the test suite
*/

const opts = {
  deployCollateral,
  collateralSpecificConstructorTests,
  collateralSpecificStatusTests,
  beforeEachRewardsTest,
  makeCollateralFixtureContext,
  mintCollateralTo,
  reduceRefPerTok,
  increaseRefPerTok,
  itClaimsRewards: it,
  itChecksTargetPerRefDefault: it,
  itCheckPriceChanges: it,
  resetFork,
  collateralName: 'CompoundV3USDC',
  chainlinkDefaultAnswer,
}

collateralTests(opts)
