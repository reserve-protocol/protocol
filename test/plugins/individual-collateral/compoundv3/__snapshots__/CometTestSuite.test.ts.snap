// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collateral: CompoundV3USDC collateral functionality Gas Reporting ERC20 transfer 1`] = `107621`;

exports[`Collateral: CompoundV3USDC collateral functionality Gas Reporting ERC20 transfer 2`] = `90521`;

exports[`Collateral: CompoundV3USDC collateral functionality Gas Reporting refresh() after full price timeout 1`] = `109052`;

exports[`Collateral: CompoundV3USDC collateral functionality Gas Reporting refresh() after full price timeout 2`] = `104315`;

exports[`Collateral: CompoundV3USDC collateral functionality Gas Reporting refresh() after hard default 1`] = `134449`;

exports[`Collateral: CompoundV3USDC collateral functionality Gas Reporting refresh() after hard default 2`] = `126766`;

exports[`Collateral: CompoundV3USDC collateral functionality Gas Reporting refresh() after oracle timeout 1`] = `109052`;

exports[`Collateral: CompoundV3USDC collateral functionality Gas Reporting refresh() after oracle timeout 2`] = `104315`;

exports[`Collateral: CompoundV3USDC collateral functionality Gas Reporting refresh() after soft default 1`] = `132572`;

exports[`Collateral: CompoundV3USDC collateral functionality Gas Reporting refresh() after soft default 2`] = `126704`;

exports[`Collateral: CompoundV3USDC collateral functionality Gas Reporting refresh() during SOUND 1`] = `126763`;

exports[`Collateral: CompoundV3USDC collateral functionality Gas Reporting refresh() during SOUND 2`] = `126763`;

exports[`Collateral: CompoundV3USDC collateral functionality Gas Reporting refresh() during soft default 1`] = `134314`;

exports[`Collateral: CompoundV3USDC collateral functionality Gas Reporting refresh() during soft default 2`] = `127045`;
